File: Source/Data/bin/cook.rb
#!/usr/bin/ruby
$KCODE="U"

if ARGV.size < 4
  $stderr.puts("usage: cook.rb phrase-freqs bpmf-mappings bpmf-base output")
  exit(1)
end

UNK_LOG_FREQ = -99.0
H_DEFLT_FREQ = -6.8

bpmf_chars = {}
bpmf_phrases = {}
phrases = {}
bpmf_phon1 = {}
bpmf_phon2 = {}
bpmf_phon3 = {}

# Reading-in a list of heterophonic words and
# its most frequent pronunciation
h1 = File.open("heterophony1.list")
while line = h1.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  key = elements[0]
  value = elements[1]
  bpmf_phon1[key] = [value]
end
h2 = File.open("heterophony2.list")
while line = h2.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  key = elements[0]
  value = elements[1]
  bpmf_phon2[key] = [value]
end
h3 = File.open("heterophony3.list")
while line = h3.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  key = elements[0]
  value = elements[1]
  bpmf_phon3[key] = [value]
end

o = File.open(ARGV[3], "w")

b = File.open(ARGV[2])
while line = b.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  
  type = elements[4]
  key = elements[0]
  value = elements[1]
  
  #if elements[4] == "big5"
    if bpmf_chars[key]
      bpmf_chars[key] += [value]
    else
      bpmf_chars[key] = [value]
    end
    
    if bpmf_phrases[key]
      bpmf_phrases[key] += [value]
    else
      bpmf_phrases[key] = [value]
    end
  #end
end

m = File.open(ARGV[1])
while line = m.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  key = elements.shift
  value = elements.join("-")
  if bpmf_phrases[key]
    bpmf_phrases[key] += [value]
  else
    bpmf_phrases[key] = [value]
  end
  #$stdout.puts("%s %d" % [key, key.length])
end

#phrases_blacklist ={}
#blacklist = File.open("blacklist.txt")
#while line = blacklist.gets
#  if line.match(/^#/)
#    next
#  end
#  line.chomp!
#  elements = line.split(/\s+/)
#  key = elements.shift
#  if key.length < 4
#    next
#  end
#  value = elements.join("-")
#  BL_LOG_FREQ = UNK_LOG_FREQ * Integer (key.length / 3 + 1) 
#  if phrases_blacklist[key]
#    phrases_blacklist[key] += [value]
#    o.puts("%s %s %d" % [key, value, BL_LOG_FREQ])
#  else
#    phrases_blacklist[key] = [value]
#    o.puts("%s %s %d" % [key, value, BL_LOG_FREQ])
#  end
#end

p = File.open(ARGV[0])
while line = p.gets
  if line.match(/^#/)
    next
  end
  line.chomp!
  elements = line.split(/\s+/)
  key = elements.shift
  #if phrases_blacklist[key]
  #  #$stdout.puts("%s %s" % [key, $stdout.puts("%s %s" % [key, phrases_blacklist[key]])])
  #  next
  #end
  value = elements.shift
  readings = bpmf_phrases[key]
  phrases[key] = true
  if readings
     # 一個字的長度目前還是 3 (unicode? or just CJK pages use length==3?)
     if key.length > 3
        readings.each do |r|
           o.puts("%s %s %s" % [key, r, value])
        end
     else
     # lookup the table from canonical list
        if bpmf_phon1[key]
           readings.each do |r|
              if bpmf_phon1[key].to_s == r
                 o.puts("%s %s %s" % [key, r, value])
              elsif bpmf_phon2[key].to_s == r
                 if value.to_f-0.28768207245178 > H_DEFLT_FREQ
                    o.puts("%s %s %f" % [key, r, value.to_f-0.28768207245178])
                 else
                    o.puts("%s %s %f" % [key, r, H_DEFLT_FREQ])
                 end
                 # l(3/4) = -0.28768207245178 / 頻率打七五折之意
              elsif bpmf_phon3[key].to_s == r
                 if value.to_f-0.28768207245178*2 > H_DEFLT_FREQ
                    o.puts("%s %s %f" % [key, r, value.to_f-0.28768207245178*2])
                 else
                    o.puts("%s %s %f" % [key, r, H_DEFLT_FREQ])
                 end
                 # l(3/4*3/4) = -0.28768207245178*2
              else
                 o.puts("%s %s %f" % [key, r, H_DEFLT_FREQ])
                 # 如果是破音字, set it to default.
                 # 很罕用的注音建議不要列入 heterophony?.list，這樣的話
                 # 就可以直接進來這個 condition
                 #$stdout.puts("%s\|%s\|" % [bpmf_phon1[key], r])
              end
           end
        else
           readings.each do |r|
              o.puts("%s %s %s" % [key, r, value])
           end
        end
     end
  end
end

bpmf_chars.each_pair do |k, v| 
  if !phrases[k]
    v.each do |r|
      o.puts("%s %s %f" % [k, r, UNK_LOG_FREQ])
    end
  end
end

File: Source/Data/bin/cook.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from cook_util import HEADER, convert_vks_rows_to_sorted_kvs_rows
import sys

__author__ = "Mengjuei Hsieh and The McBopomofo Authors"
__copyright__ = "Copyright 2012 and onwards The McBopomofo Authors"
__license__ = "MIT"

UNK_LOG_FREQ = -99.0
H_DEFLT_FREQ = -6.8

bpmf_chars = {}
bpmf_phrases = {}
phrases = {}
bpmf_phon1 = {}
bpmf_phon2 = {}
bpmf_phon3 = {}
bpmf_hetero = {}


if __name__ == '__main__':
    """bin/cook.py PhraseFreq.txt BPMFMappings.txt BPMFBase.txt data.txt"""
    if len(sys.argv) < 7:
        sys.exit('Usage: cook.py phrase-freqs bpmf-mappings bpmf-base punctuations symbols output')
    #  Read a list of heterophonic singulars and its estimated frequency
    #  not active yet
    # try:
    #     handle = open('heterophony.list', "r")
    # except IOError as e:
    #     print("({})".format(e))
    # while True:
    #     line = handle.readline()
    #     if not line: break
    #     if line[0] == '#': continue
    #     elements = line.rstrip().split()
    #     myword = elements[0]
    #     myvalue = {elements[1], elements[2]}
    #     if myword in bpmf_hetero:
    #         bpmf_hetero[myword].append(myvalue)
    #     else:
    #         bpmf_hetero[myword] = []
    #         bpmf_hetero[myword].append(myvalue)
    # handle.close()
    #  Reading-in a list of heterophonic words and
    #  its most frequent pronunciation
    try:
        handle = open('heterophony1.list', "r")
    except IOError as e:
        print("({})".format(e))
    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        # if elements[0] in bpmf_hetero: break
        bpmf_phon1[elements[0]] = elements[1]
    handle.close()
    try:
        handle = open('heterophony2.list', "r")
    except IOError as e:
        print("({})".format(e))
    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        # if elements[0] in bpmf_hetero: break
        bpmf_phon2[elements[0]] = elements[1]
    handle.close()
    try:
        handle = open('heterophony3.list', "r")
    except IOError as e:
        print("({})".format(e))
    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        # if elements[0] in bpmf_hetero: break
        bpmf_phon3[elements[0]] = elements[1]
    handle.close()
    # bpmfbase
    handle = open(sys.argv[3], "r")
    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        mytype = elements[4]
        mykey = elements[0]
        myvalue = elements[1]
        # print mykey
        if mykey in bpmf_chars:
            bpmf_chars[mykey].append(myvalue)
        else:
            bpmf_chars[mykey] = []
            bpmf_chars[mykey].append(myvalue)
        if mykey in bpmf_phrases:
            bpmf_phrases[mykey].append(myvalue)
        else:
            bpmf_phrases[mykey] = []
            bpmf_phrases[mykey].append(myvalue)
    handle.close()
    # bpmf-mappings
    handle = open(sys.argv[2], "r")
    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        mykey = elements.pop(0)
        myvalue = "-".join(elements)
        # print mykey
        # print myvalue
        if mykey in bpmf_phrases:
            bpmf_phrases[mykey].append(myvalue)
        else:
            bpmf_phrases[mykey] = []
            bpmf_phrases[mykey].append(myvalue)
    handle.close()
    # phrase-freqs
    handle = open(sys.argv[1], "r")

    output = []

    while True:
        line = handle.readline()
        if not line: break
        if line[0] == '#': continue
        elements = line.rstrip().split()
        mykey = elements.pop(0)
        myvalue = elements.pop(0)
        try:
            readings = bpmf_phrases[mykey]
        except:
            sys.exit('[ERROR] %s key mismatches.' % mykey)
        phrases[mykey] = True
        # print mykey
        if readings:
            # 剛好一個中文字字的長度目前還是 3 (標點、聲調好像都是2)
            if len(mykey) > 3:
                for r in readings:
                    output.append((mykey, r, myvalue))
                    pass
                continue
            else:
                # lookup the table from canonical list
                for r in readings:
                    if mykey not in bpmf_phon1:
                        output.append((mykey, r, myvalue))
                        continue
                    elif str(bpmf_phon1[mykey]) == r:
                        output.append((mykey, r, myvalue))
                        continue
                    elif mykey not in bpmf_phon2:
                        output.append((mykey, r, H_DEFLT_FREQ))
                        continue
                    elif str(bpmf_phon2[mykey]) == r:
                        # l(3/4) = -0.28768207245178 / 頻率打七五折之意
                        # l(1/2) = -0.69314718055994 / 頻率打五折之意
                        if float(myvalue) - 0.69314718055994 > H_DEFLT_FREQ:
                            output.append((mykey, r, float(myvalue) - 0.69314718055994))
                            continue
                        else:
                            output.append((mykey, r, H_DEFLT_FREQ))
                            continue
                    elif mykey not in bpmf_phon3:
                        output.append((mykey, r, H_DEFLT_FREQ))
                        continue
                    elif str(bpmf_phon3[mykey]) == r:
                        # l(3/4*3/4) = -0.28768207245178*2
                        # l(1/2*1/2) = -0.69314718055994*2
                        if float(myvalue) - 0.69314718055994 * 2 > H_DEFLT_FREQ:
                            output.append((mykey, r, float(myvalue) - 0.69314718055994 * 2))
                            continue
                        else:
                            output.append((mykey, r, H_DEFLT_FREQ))
                            continue
                    output.append((mykey, r, H_DEFLT_FREQ))
                    # 如果是破音字, set it to default.
                    # 很罕用的注音建議不要列入 heterophony?.list，這樣的話
                    # 就可以直接進來這個 condition
    handle.close()
    for k in bpmf_chars:
        if k not in phrases:
            for v in bpmf_chars[k]:
                output.append((k, v, UNK_LOG_FREQ))
                pass

    with open(sys.argv[4]) as punctuations_file:
        for line in punctuations_file:
            row = line.rstrip().split(" ")
            assert len(row) == 3
            output.append(tuple(row))

    with open(sys.argv[5]) as symbols_file:
        for line in symbols_file:
            row = line.rstrip().split(" ")
            assert len(row) == 3, row
            output.append(tuple(row))

    with open(sys.argv[6]) as macro_file:
        for line in macro_file:
            row = line.rstrip().split(" ")
            assert len(row) == 3, row
            output.append(tuple(row))

    output = convert_vks_rows_to_sorted_kvs_rows(output)
    with open(sys.argv[-1], "w") as fout:
        fout.write(HEADER)
        for row in output:
            if type(row[-1]) is float:
                fout.write('%s %s %f\n' % row)
            else:
                fout.write('%s %s %s\n' % row)
